In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [2]:
df = pd.read_csv("owid-covid-data.csv")
df

iso_code continent     location        date  total_cases  new_cases  \
0           AFG      Asia  Afghanistan  2020-02-24          5.0        5.0   
1           AFG      Asia  Afghanistan  2020-02-25          5.0        0.0   
2           AFG      Asia  Afghanistan  2020-02-26          5.0        0.0   
3           AFG      Asia  Afghanistan  2020-02-27          5.0        0.0   
4           AFG      Asia  Afghanistan  2020-02-28          5.0        0.0   
...         ...       ...          ...         ...          ...        ...   
166321      ZWE    Africa     Zimbabwe  2022-03-01     236871.0      491.0   
166322      ZWE    Africa     Zimbabwe  2022-03-02     237503.0      632.0   
166323      ZWE    Africa     Zimbabwe  2022-03-03     237503.0        0.0   
166324      ZWE    Africa     Zimbabwe  2022-03-04     238739.0     1236.0   
166325      ZWE    Africa     Zimbabwe  2022-03-05     239019.0      280.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
0                      NaN           NaN         NaN                  NaN   
1                      NaN           NaN         NaN                  NaN   
2                      NaN           NaN         NaN                  NaN   
3                      NaN           NaN         NaN                  NaN   
4                      NaN           NaN         NaN                  NaN   
...                    ...           ...         ...                  ...   
166321             413.000        5395.0         0.0                1.000   
166322             416.286        5396.0         1.0                1.143   
166323             362.286        5396.0         0.0                0.857   
166324             467.429        5397.0         1.0                0.714   
166325             459.429        5397.0         0.0                0.571   

        ...  female_smokers  male_smokers  handwashing_facilities  \
0       ...             NaN           NaN                  37.746   
1       ...             NaN           NaN                  37.746   
2       ...             NaN           NaN                  37.746   
3       ...             NaN           NaN                  37.746   
4       ...             NaN           NaN                  37.746   
...     ...             ...           ...                     ...   
166321  ...             1.6          30.7                  36.791   
166322  ...             1.6          30.7                  36.791   
166323  ...             1.6          30.7                  36.791   
166324  ...             1.6          30.7                  36.791   
166325  ...             1.6          30.7                  36.791   

        hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                              0.5            64.83                    0.511   
1                              0.5            64.83                    0.511   
2                              0.5            64.83                    0.511   
3                              0.5            64.83                    0.511   
4                              0.5            64.83                    0.511   
...                            ...              ...                      ...   
166321                         1.7            61.49                    0.571   
166322                         1.7            61.49                    0.571   
166323                         1.7            61.49                    0.571   
166324                         1.7            61.49                    0.571   
166325                         1.7            61.49                    0.571   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                        NaN                          NaN   
1                                        NaN                          NaN   
2                                        NaN                          NaN   
3                                        NaN   

In [3]:
USA_data = df[df['location'] == 'United States']
USA_data = USA_data.set_index('date')
Brazil_data = df[df['location'] == 'Brazil']
Brazil_data = Brazil_data.set_index('date')
Germany_data = df[df['location'] == 'Germany']
Germany_data = Germany_data.set_index('date')
China_data = df[df['location'] == 'China']
China_data = China_data.set_index('date')
Nigeria_data = df[df['location'] == 'Nigeria']
Nigeria_data = Nigeria_data.set_index('date')
Australia_data = df[df['location'] == 'Australia']
Australia_data = Australia_data.set_index('date')

In [4]:
USA_train = USA_data[0:710]
USA_test = USA_data[711:800]
Brazil_train = Brazil_data[0:710]
Brazil_test = Brazil_data[711:800]
Germany_train = Germany_data[0:710]
Germany_test = Germany_data[711:800]
China_train = China_data[0:710]
China_test = China_data[711:800]
Nigeria_train = Nigeria_data[0:710]
Nigeria_test = Nigeria_data[711:800]
Australia_train = Australia_data[0:710]
Australia_test = Australia_data[711:800]

In [5]:
df.columns

Index(['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases',
       'new_cases_smoothed', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_cases_per_million',
       'new_cases_per_million', 'new_cases_smoothed_per_million',
       'total_deaths_per_million', 'new_deaths_per_million',
       'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients',
       'icu_patients_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'weekly_icu_admissions',
       'weekly_icu_admissions_per_million', 'weekly_hosp_admissions',
       'weekly_hosp_admissions_per_million', 'new_tests', 'total_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations',
       'people_vaccinated', 'people_fully_vaccinated', 'total_boosters',
       'new_vaccinations', 'new_vaccinations_smoothed',
       't

# Vaccinations

Many countries are missing vaccination data only within an initial time period (about a 1 year delay), but are generally complete going forward. Each of the countries will be check to verify this information, and may be chosen as predictors despite higher missingness proportions. We anticipate our models to be robust to this specific pattern of missingness.

# USA

In [34]:
missing_USA = pd.DataFrame(USA_train.isnull().sum())
missing_USA['% missing'] = USA_train.isnull().sum()/USA_train.shape[0]
missing_USA.columns = ["Missing Count", "Proportion Missing"]
missing_USA = missing_USA.sort_values(by="Proportion Missing", ascending=True)
missing_USA = missing_USA[missing_USA["Proportion Missing"] < .50]
print(missing_USA.shape)
missing_USA

(57, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
population                                              0            0.000000
new_cases_per_million                                   1            0.001408
new_cases                                               1            0.001408
new_cases_smoothed_per_million                          7            0.009859
new_cases_smoothed                                      7            0.009859
total_deaths_per_million                               38            0.053521
new_deaths                                             38            0.053521
total_deaths                                           38            0.053521
new_deaths_per_million                                 38            0.053521
tests_units                                            39            0.054930
total_tests_per_thousand                               39            0.054930
new_tests                                              39            0.054930
total_tests                                            39            0.054930
new_tests_per_thousand                                 39            0.054930
reproduction_rate                                      43            0.060563
new_deaths_smoothed                                    44            0.061972
new_deaths_smoothed_per_million                        44            0.061972
positive_rate                                          45            0.063380
tests_per_case                                         45            0.063380
new_tests_smoothed                                     46            0.064789
new_tests_smoothed_per_thousand                        46            0.064789
icu_patients                                          175            0.246479
icu_patients_per_million                              175            0.246479
hosp_patients                                         175            0.246479
hosp_patients_per_million                             175            0.246479
weekly_hosp_admissions_per_million                    181            0.254930
weekly_hosp_admissions                                181            0.254930
total_vaccinations_per_hundred                        326            0.459155
people_vaccinated                                     326            0.459155
people_fully_vaccinated                               326            0.459155
people_vaccinated_per_hundred                         326            0.459155
people_fully_vaccinated_per_hundred                   326         

In [45]:
# Selecting predictors with adequate missingness

selected_USA = missing_USA.index
USA_train = USA_train[selected_USA] 
USA_test = USA_test[selected_USA] 

# Brazil

In [31]:
missing_Brazil = pd.DataFrame(Brazil_train.isnull().sum())
missing_Brazil['% missing'] = Brazil_train.isnull().sum()/Brazil_train.shape[0]
missing_Brazil.columns = ["Missing Count", "Proportion Missing"]
missing_Brazil = missing_Brazil.sort_values(by="Proportion Missing", ascending=True)
missing_Brazil = missing_Brazil[missing_Brazil["Proportion Missing"] < .50]
print(missing_Brazil.shape)
missing_Brazil

(47, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
human_development_index                                 0            0.000000
population                                              0            0.000000
stringency_index                                        0            0.000000
total_cases_per_million                                 0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
new_cases_per_million                                   2            0.002817
new_cases                                               2            0.002817
reproduction_rate                                      17            0.023944
new_deaths_per_million                                 20            0.028169
total_deaths_per_million                               20            0.028169
new_cases_smoothed_per_million                         20            0.028169
new_deaths                                             20            0.028169
total_deaths                                           20            0.028169
new_cases_smoothed                                     20            0.028169
new_deaths_smoothed                                    26            0.036620
new_deaths_smoothed_per_million                        26            0.036620
tests_units                                            60            0.084507
new_tests_smoothed                                     67            0.094366
new_tests_smoothed_per_thousand                        67            0.094366
new_vaccinations_smoothed                             327            0.460563
new_vaccinations_smoothed_per_million                 327            0.460563
new_people_vaccinated_smoothed                        327            0.460563
new_people_vaccinated_smoothed_per_hundred            327            0.460563
total_vaccinations                                    332            0.467606
total_vaccinations_per_hundred                        332            0.467606
new_vaccinations                                      335            0.471831
people_vaccinated                                     336            0.473239
people_vaccinated_per_hundred                         336            0.473239
people_fully_vaccinated                               352            0.495775
people_fully_vaccinated_per_hundred                   352            0.495775
total_tests_per_thousand                              353            0.497183
total_tests                                           353            0.497183

In [44]:
# Selecting predictors with adequate missingness

selected_Brazil = missing_Brazil.index[:-2]
Brazil_train = Brazil_train[selected_Brazil] 
Brazil_test = Brazil_test[selected_Brazil] 

# Germany

In [48]:
missing_Germany = pd.DataFrame(Germany_train.isnull().sum())
missing_Germany['% missing'] = Germany_train.isnull().sum()/Germany_train.shape[0]
missing_Germany.columns = ["Missing Count", "Proportion Missing"]
missing_Germany = missing_Germany.sort_values(by="Proportion Missing", ascending=True)
missing_Germany = missing_Germany[missing_Germany["Proportion Missing"] < .50]
print(missing_Germany.shape)
missing_Germany

(50, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
population                                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
continent                                               0            0.000000
location                                                0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
total_cases                                             0            0.000000
new_cases_per_million                                   2            0.002817
new_cases                                               2            0.002817
new_cases_smoothed                                     20            0.028169
new_cases_smoothed_per_million                         20            0.028169
weekly_hosp_admissions                                 34            0.047887
weekly_hosp_admissions_per_million                     34            0.047887
reproduction_rate                                      35            0.049296
total_deaths                                           42            0.059155
total_deaths_per_million                               42            0.059155
new_deaths                                             46            0.064789
new_deaths_per_million                                 46            0.064789
tests_units                                            48            0.067606
icu_patients_per_million                               53            0.074648
icu_patients                                           53            0.074648
new_tests_smoothed                                     55            0.077465
new_tests_smoothed_per_thousand                        55            0.077465
new_deaths_smoothed_per_million                        76            0.107042
new_deaths_smoothed                                    76            0.107042
total_vaccinations                                    335            0.471831
people_vaccinated                                     335            0.471831
people_fully_vaccinated                               335            0.471831
total_vaccinations_per_hundred                        335            0.471831
people_vaccinated_per_hundred                         335            0.471831
people_fully_vaccinated_per_hundred                   335            0.471831
new_people_vaccinated_smoothed_per_hundred            336            0.473239
new_vaccinations_smoothed_per_million                 336            0.473239
new_vaccinations_smoothed                             336            0.473239
new_vaccinations                                      336            0.473239
new_people_vaccinated_smoothed                        336            0.473239
total_boosters_per_hundred                            337            0.474648
total_boosters                                        337            0.474648

In [49]:
# Selecting predictors with adequate missingness

selected_Germany = missing_Germany.index
Germany_train = Germany_train[selected_Germany] 
Germany_test = Germany_test[selected_Germany] 

# China

In [52]:
missing_China = pd.DataFrame(China_train.isnull().sum())
missing_China['% missing'] = China_train.isnull().sum()/China_train.shape[0]
missing_China.columns = ["Missing Count", "Proportion Missing"]
missing_China = missing_China.sort_values(by="Proportion Missing", ascending=True)
missing_China = missing_China[missing_China["Proportion Missing"] < .65]
print(missing_China.shape)
missing_China

(36, 2)


Missing Count  Proportion Missing
iso_code                                           0            0.000000
stringency_index                                   0            0.000000
population                                         0            0.000000
population_density                                 0            0.000000
median_age                                         0            0.000000
aged_65_older                                      0            0.000000
aged_70_older                                      0            0.000000
gdp_per_capita                                     0            0.000000
extreme_poverty                                    0            0.000000
cardiovasc_death_rate                              0            0.000000
diabetes_prevalence                                0            0.000000
female_smokers                                     0            0.000000
male_smokers                                       0            0.000000
life_expectancy                                    0            0.000000
human_development_index                            0            0.000000
hospital_beds_per_thousand                         0            0.000000
continent                                          0            0.000000
total_deaths_per_million                           0            0.000000
location                                           0            0.000000
total_cases_per_million                            0            0.000000
total_deaths                                       0            0.000000
total_cases                                        0            0.000000
reproduction_rate                                  1            0.001408
new_deaths_per_million                             2            0.002817
new_deaths                                         2            0.002817
new_cases_per_million                              3            0.004225
new_cases                                          3            0.004225
new_deaths_smoothed_per_million                   14            0.019718
new_deaths_smoothed                               14            0.019718
new_cases_smoothed                                21            0.029577
new_cases_smoothed_per_million                    21            0.029577
new_vaccinations_smoothed                        329            0.463380
new_vaccinations_smoothed_per_million            329            0.463380
total_vaccinations                               413            0.581690
total_vaccinations_per_hundred                   413            0.581690
new_vaccinations                                 426            0.600000

In [55]:
# Selecting predictors with adequate missingness

# -3 is used to remove the tail-end on the previous dataframe, whose missingness is about 50%. 
# Make sure to change this number if you change the threshold or decide to include more/less predictors from the dataframe.

selected_China = missing_China.index[:-3]
China_train = China_train[selected_China] 
China_test = China_test[selected_China] 

# Nigeria

In [58]:
missing_Nigeria = pd.DataFrame(Nigeria_train.isnull().sum())
missing_Nigeria['% missing'] = Nigeria_train.isnull().sum()/Nigeria_train.shape[0]
missing_Nigeria.columns = ["Missing Count", "Proportion Missing"]
missing_Nigeria = missing_Nigeria.sort_values(by="Proportion Missing", ascending=True)
missing_Nigeria = missing_Nigeria[missing_Nigeria["Proportion Missing"] < .60]
print(missing_Nigeria.shape)
missing_Nigeria

(39, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
handwashing_facilities                                  0            0.000000
life_expectancy                                         0            0.000000
human_development_index                                 0            0.000000
population                                              0            0.000000
stringency_index                                        0            0.000000
new_cases_per_million                                   0            0.000000
total_cases_per_million                                 0            0.000000
new_cases                                               0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
new_cases_smoothed_per_million                          6            0.008451
new_cases_smoothed                                      6            0.008451
total_deaths_per_million                               24            0.033803
total_deaths                                           24            0.033803
new_deaths                                             25            0.035211
new_deaths_per_million                                 25            0.035211
reproduction_rate                                      31            0.043662
new_deaths_smoothed                                    37            0.052113
new_deaths_smoothed_per_million                        37            0.052113
tests_units                                            40            0.056338
tests_per_case                                         47            0.066197
positive_rate                                          47            0.066197
new_tests_smoothed                                     47            0.066197
new_tests_smoothed_per_thousand                        47            0.066197
new_vaccinations_smoothed                             371            0.522535
new_people_vaccinated_smoothed_per_hundred            371            0.522535
new_people_vaccinated_smoothed                        371            0.522535
new_vaccinations_smoothed_per_million                 371            0.522535

In [59]:
# Selecting predictors with adequate missingness

selected_Nigeria = missing_Nigeria.index[:-4]
Nigeria_train = Nigeria_train[selected_Nigeria] 
Nigeria_test = Nigeria_test[selected_Nigeria] 

# Australia

In [62]:
missing_Australia = pd.DataFrame(Australia_train.isnull().sum())
missing_Australia['% missing'] = Australia_train.isnull().sum()/Australia_train.shape[0]
missing_Australia.columns = ["Missing Count", "Proportion Missing"]
missing_Australia = missing_Australia.sort_values(by="Proportion Missing", ascending=True)
missing_Australia = missing_Australia[missing_Australia["Proportion Missing"] < .60]
print(missing_Australia.shape)
missing_Australia

(55, 2)


Missing Count  Proportion Missing
iso_code                                                0            0.000000
population_density                                      0            0.000000
median_age                                              0            0.000000
aged_65_older                                           0            0.000000
aged_70_older                                           0            0.000000
gdp_per_capita                                          0            0.000000
extreme_poverty                                         0            0.000000
cardiovasc_death_rate                                   0            0.000000
diabetes_prevalence                                     0            0.000000
female_smokers                                          0            0.000000
male_smokers                                            0            0.000000
hospital_beds_per_thousand                              0            0.000000
life_expectancy                                         0            0.000000
stringency_index                                        0            0.000000
new_cases_per_million                                   0            0.000000
total_cases_per_million                                 0            0.000000
human_development_index                                 0            0.000000
new_cases                                               0            0.000000
total_cases                                             0            0.000000
location                                                0            0.000000
continent                                               0            0.000000
population                                              0            0.000000
new_cases_smoothed_per_million                          6            0.008451
new_cases_smoothed                                      6            0.008451
total_deaths_per_million                               35            0.049296
total_deaths                                           35            0.049296
new_deaths                                             38            0.053521
new_deaths_per_million                                 38            0.053521
tests_units                                            41            0.057746
total_tests_per_thousand                               41            0.057746
total_tests                                            41            0.057746
new_tests                                              42            0.059155
new_tests_per_thousand                                 42            0.059155
reproduction_rate                                      45            0.063380
new_tests_smoothed                                     48            0.067606
new_tests_smoothed_per_thousand                        48            0.067606
positive_rate                                          48            0.067606
tests_per_case                                         48            0.067606
new_deaths_smoothed_per_million                        62            0.087324
new_deaths_smoothed                                    62            0.087324
icu_patients_per_million                               65            0.091549
hosp_patients                                          65            0.091549
hosp_patients_per_million                              65            0.091549
icu_patients                                           65            0.091549
new_vaccinations_smoothed                             393            0.553521
new_vaccinations_smoothed_per_million                 393            0.553521
new_people_vaccinated_smoothed                        393            0.553521
new_people_vaccinated_smoothed_per_hundred            393            0.553521
total_vaccinations                                    397            0.559155
people_vaccinated                                     397            0.559155
people_vaccinated_per_hundred                         397         

In [63]:
# Selecting predictors with adequate missingness

selected_Australia = missing_Australia.index[:-11]
Australia_train = Australia_train[selected_Australia] 
Australia_test = Australia_test[selected_Australia] 

# Notes, Selection Justification

* Though we are checking predictor missingness off of the training data set, we are looking at the actual data spread from the entire CSV file in excel. This way, we can check the date ranges the missingness falls within, if there is a pattern.

* Final predictor count for each country is 1 less than listed from the missingness data frames generated, because we are exclusing iso_code as a predictor.


1. USA_train has 45 predictors below 30% missingness. The next 11 predictors had missingness between 45-46%. However, these predictors displayed missingness only up until middle of December 2020, so they are kept. Thus, 56 predictors are kept for USA, demonstrating up to 46% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 25%. (56)

2. Brazil_train only has 33 predictors below 30% missingness. The next 13 predictors had missingness between 46-49%. However, 11 of these predictors displayed missingness only up until early February of 2021, so they are kept. The other two predictors (total_tests and total_tests_per_thousand) had sporadic missingness all across the timeline, so they were removed. Thus, 44 predictors are kept for Brazil, demonstrating upto 47.3% missingness. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 9.4%. (44)

3. Germany_train has only 36 predictors below 30% missingness. The next 13 predictors had missingness around 47%. However, these predictors only displayed missingess up until the end of December 2020, so they are kept. Thus, 49 predictors are kept for Germany, demonstrating missingness up to 47%. There is a large gap between the additional set of predictors and the original set, whose highest missingness was around 10%. (49)

4. China_train only has 30 predictors below 30% missingness. The next 2 are new_vaccinations_smoothed and new_vaccinations_smoothed_per_million, which had missingness around 46%. However, these two only displayed missingness up until the middle of December 2020, so they are kept. All of the other vaccination based predictors had missingness about 50%, so they are not pursued at the moment. Thus, a total of 32 predictors are selected. (32)

5. Nigeria_train has only 34 predictors below 30% missingness. The predictors following these display missingness above 50%, so they are not pursued. Thus, we will select the original 34 predictors, which display missingness up to around 6%. (34)

6. Australia_train has 43 predictors below 30% missingness. Tge predictors following these display missingness above 50%, so they are not pursued. Thus, we will select the original 43 predictors, which display missingness up to around 9%. (43)


* Questions/Concerns:

1. Should we include all vaccine related predictors anyway, since they logically contribute to new cases? 
2. Should we include some of these predictors up to 60%?
3. (Should I calculate missingness off of train or the whole set? I felt missingness impacted training more anyways, and testing demonstrated very minimal missigness. This is an easy code fix.)
4. China and Nigeria are the only concerning countries, since they are only around 1/2 of predictors while the others are at 2/3.
5. Australia is only mildly concerning, as it has sufficient predictors but lacks all of the vaccine-based predictors.